File: Java/BinarySearch/BS20 - Minimise Maximum Distance between Gas Stations.java
package BinarySearch;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

class Utility {
    public static <T> List<T> getList(Integer size, T defaultValue) {
        List<T> list = new ArrayList<>();
        for (int i = 0; i < size; i += 1) {
            list.add(defaultValue);
        }
        return list;
    }
}

class LinearSolution {
    public static Double getMinimizedMaxDistance(List<Integer> arr, Integer k) {
        /**
         * Time complexity is O(nk) and space complexity is O(n).
         */

        int n = arr.size();
        
        // get the list of slots in O(n) time and O(n) space.
        List<Integer> slots = Utility.getList(n - 1, 0);
        
        // loop on the `k` gas stations that needs to be placed in k iterations.
        for (int i = 0; i < k; i += 1) {
            // store the max slot length and index based on the current state of the slots.
            int maxSlotIndex = -1;
            double maxSlotLength = 0.0;
            
            // loop on the existing gas stations in `n - 1` iterations.
            for (int j = 0; j < n - 1; j += 1) {
                // compute the slot length of each slot and update the max slot length and index. Basically, this inner
                // loop is trying to find the maximum slot where the next gas station can be placed.
                int left = arr.get(j);
                int right = arr.get(j + 1);
                double slotLength = (right - left)/((double) slots.get(j) + 1.0);
                if (maxSlotLength < slotLength) {
                    maxSlotLength = slotLength;
                    maxSlotIndex = j;
                }
            }

            // once the largest slot is found, put the next gas station there.
            slots.set(maxSlotIndex, slots.get(maxSlotIndex) + 1);
        }

        // finally, again find the max slot length which would point to the minimized maximum length between any two gas
        // stations. This is our final answer and this takes O(n) time.
        double maxSlotLength = 0;
        for (int j = 0; j < n - 1; j += 1) {
            int left = arr.get(j);
            int right = arr.get(j + 1);
            double slotLength = (right - left)/((double) slots.get(j) + 1.0);
            if (maxSlotLength < slotLength) {
                maxSlotLength = slotLength;
            }
        }
        return maxSlotLength;
    }
}


class Slot implements Comparable<Slot> {
    private Double length;
    private Integer gasStationsPlaced;

    public Slot(Double length, Integer gasStationsPlaced) {
        this.length = length;
        this.gasStationsPlaced = gasStationsPlaced;
    }

    public Double getLength() {
        return length;
    }

    public void setLength(Double length) {
        this.length = length;
    }

    public Integer getGasStationsPlaced() {
        return gasStationsPlaced;
    }

    public void setGasStationsPlaced(Integer gasStationsPlaced) {
        this.gasStationsPlaced = gasStationsPlaced;
    }


    @Override
    public int compareTo(Slot o) {
        if (this.length - o.getLength() < 0) {
            return -1;
        } else if (this.length - o.getLength()  == 0) {
            return 0;
        } else {
            return  1;
        }
    }

    @Override
    public String toString() {
        return String.format("Slot length = %s, Gas stations placed = %d", getLength(), getGasStationsPlaced());
    }
}

class MaxHeap<T extends Comparable<T>> {
    private List<T> heap;

    public MaxHeap() {
        this.heap = new ArrayList<>();
    }

    public boolean isEmpty() {
        return heap.isEmpty();
    }

    public Integer getLci(int pi) {
        int lci = 2*pi + 1;
        return 0 <= lci && lci < heap.size() ? lci : null;
    }

    public Integer getRci(int pi) {
        int rci = 2*pi + 2;
        return 0 <= rci && rci < heap.size() ? rci : null;
    }

    public Integer getPi(int ci) {
        if (ci == 0) {
            return null;
        }
        int pi = (ci - 1)/2;
        return 0 <= pi && pi < heap.size() ? pi : null;
    }

    public Integer getMaxChildIndex(Integer lci, Integer rci) {
        if (lci == null && rci == null) {
            return null;
        }
        if (lci == null) {
            return rci;
        }
        if (rci == null) {
            return lci;
        }
        Integer maxChildIndex = lci;
        if (heap.get(rci).compareTo(heap.get(maxChildIndex)) > 0) {
            maxChildIndex = rci;
        }
        return maxChildIndex;
    }

    public void maxHeapifyUp(Integer startIndex) {
        if (startIndex.equals(0)) {
            return;
        }
        Integer pi = getPi(startIndex);
        Integer lci = getLci(pi);
        Integer rci = getRci(pi);
        Integer maxChildIndex = getMaxChildIndex(lci, rci);
        if (maxChildIndex != null) {
            if (heap.get(pi).compareTo(heap.get(maxChildIndex)) < 0) {
                Collections.swap(heap, pi, maxChildIndex);
            }
            maxHeapifyUp(pi);
        }
    }

    public void maxHeapifyDown(Integer pi) {
        Integer lci = getLci(pi);
        Integer rci = getRci(pi);
        Integer maxChildIndex = getMaxChildIndex(lci, rci);
        if (maxChildIndex != null) {
            if (heap.get(pi).compareTo(heap.get(maxChildIndex)) < 0) {
                Collections.swap(heap, pi, maxChildIndex);
            }
            maxHeapifyDown(maxChildIndex);
        }
    }

    public void insert(T x) {
        heap.add(x);
        maxHeapifyUp(heap.size() - 1);
    }

    public T pop() {
        if (isEmpty()) {
            return null;
        }
        T item = heap.getFirst();
        Collections.swap(heap, 0, heap.size() - 1);
        heap.removeLast();
        maxHeapifyDown(0);
        return item;
    }
}

class BetterSolution {
    public static Double getMinimizedMaxDistance(List<Integer> arr, Integer k) {
        MaxHeap<Slot> maxHeap = new MaxHeap<>();
        List<Slot> slots = new ArrayList<>();
        for (int i = 0; i < arr.size() - 1; i += 1) {
            slots.add(new Slot((double) (arr.get(i + 1) - arr.get(i)), 0));
        }
        slots.forEach(maxHeap::insert);
        for (int i = 0; i < k; i += 1) {
            Slot maxSlot = maxHeap.pop();
            double length = maxSlot.getLength();
            int gasStationsPlaced = maxSlot.getGasStationsPlaced();
            double slotLength = length * (gasStationsPlaced + 1);
            double newSlotLength = slotLength/(gasStationsPlaced + 1.0 + 1.0);
            maxHeap.insert(new Slot(newSlotLength, gasStationsPlaced + 1));
        }
        Slot minimizedMaxSlot = maxHeap.pop();
        return minimizedMaxSlot.getLength();
    }
}


class Solution {
    public static void main(String[] args) {
        System.out.println(LinearSolution.getMinimizedMaxDistance(Arrays.asList(1, 13, 17, 23), 5));
        System.out.println(LinearSolution.getMinimizedMaxDistance(Arrays.asList(1,2,3,4,5,6,7), 6));
        System.out.println(LinearSolution.getMinimizedMaxDistance(Arrays.asList(1, 2, 3, 4, 5), 4));
        System.out.println(LinearSolution.getMinimizedMaxDistance(IntStream.range(1, 11).boxed().toList(), 1));
        System.out.println(LinearSolution.getMinimizedMaxDistance(Arrays.asList(3, 6, 12, 19, 33, 44, 67, 72, 89, 95), 2));
        System.out.println(LinearSolution.getMinimizedMaxDistance(Arrays.asList(1, 13, 17, 23), 5));

        System.out.println();
        System.out.println(BetterSolution.getMinimizedMaxDistance(Arrays.asList(1, 13, 17, 23), 5));
        System.out.println(BetterSolution.getMinimizedMaxDistance(Arrays.asList(1,2,3,4,5,6,7), 6));
        System.out.println(BetterSolution.getMinimizedMaxDistance(Arrays.asList(1, 2, 3, 4, 5), 4));
        System.out.println(BetterSolution.getMinimizedMaxDistance(IntStream.range(1, 11).boxed().toList(), 1));
        System.out.println(BetterSolution.getMinimizedMaxDistance(Arrays.asList(3, 6, 12, 19, 33, 44, 67, 72, 89, 95), 2));
        System.out.println(BetterSolution.getMinimizedMaxDistance(Arrays.asList(1, 13, 17, 23), 5));
    }
}

File: Python/Practice Set 4/Binary Search/BS20 - Minimise Maximum Distance between Gas Stations.py
# Problem link - https://www.geeksforgeeks.org/problems/minimize-max-distance-to-gas-station/1
# Solution - https://www.youtube.com/watch?v=kMSBvlZ-_HA&list=PLgUwDviBIf0pMFMWuuvDNMAkoQFi-h0ZF&index=21


class Utility:
    @staticmethod
    def swap(arr, i, j):
        arr[i], arr[j] = arr[j], arr[i]


class MaxHeap:
    def __init__(self):
        self.heap = []

    def is_empty(self):
        return len(self.heap) == 0

    def get_lci(self, pi):
        lci = 2 * pi + 1
        return lci if lci in range(len(self.heap)) else None

    def get_rci(self, pi):
        rci = 2 * pi + 2
        return rci if rci in range(len(self.heap)) else None

    def get_pi(self, ci):
        if ci == 0:
            return
        pi = int((ci - 1) / 2)
        return pi if pi in range(len(self.heap)) else None

    def get_max_child_index(self, lci, rci):
        if lci is None and rci is None:
            return
        if lci is None:
            return rci
        if rci is None:
            return lci
        max_child_index = lci
        if self.heap[rci] > self.heap[max_child_index]:
            max_child_index = rci
        return max_child_index

    def heapify_up(self, start_index):
        if start_index == 0:
            return
        pi = self.get_pi(start_index)
        lci, rci = self.get_lci(pi), self.get_rci(pi)
        max_child_index = self.get_max_child_index(lci, rci)
        if max_child_index is not None:
            if self.heap[pi] < self.heap[max_child_index]:
                Utility.swap(self.heap, pi, max_child_index)
            self.heapify_up(pi)

    def heapify_down(self, pi):
        lci, rci = self.get_lci(pi), self.get_rci(pi)
        max_child_index = self.get_max_child_index(lci, rci)
        if max_child_index is not None:
            if self.heap[pi] < self.heap[max_child_index]:
                Utility.swap(self.heap, pi, max_child_index)
            self.heapify_down(max_child_index)

    def insert(self, x):
        self.heap.append(x)
        self.heapify_up(len(self.heap) - 1)

    def pop(self):
        if self.is_empty():
            return
        item = self.heap[0]
        Utility.swap(self.heap, 0, -1)
        del self.heap[-1]
        self.heapify_down(0)
        return item


class Element:
    def __init__(self, distance, index, placed_count):
        self.d = distance
        self.i = index
        self.p = placed_count

    def __lt__(self, other):
        if self.d == other.d:
            return self.i < other.i
        return self.d < other.d

    def __gt__(self, other):
        if self.d == other.d:
            return self.i > other.i
        return self.d > other.d


class Solution:
    @staticmethod
    def place_gas_stations(arr, k):
        """
            Time complexity is O({n + k} * log(n)) and space complexity is O(n).
        """
        if k <= 0:
            return
        pq = MaxHeap()

        # takes O(n * log(n)) time
        Solution._initialize_pq(pq, arr)

        # takes O(k * log(n)) time
        while k > 0:
            element = pq.pop()
            distance, index, placed_count = element.d, element.i, element.p
            new_distance = distance * (placed_count + 1) / (placed_count + 2)
            pq.insert(Element(new_distance, index, placed_count + 1))
            k -= 1
        return pq.pop().d

    @staticmethod
    def _initialize_pq(pq, arr):
        for i in range(len(arr) - 1):
            element = Element(arr[i + 1] - arr[i], i, 0)
            pq.insert(element)


print(Solution.place_gas_stations([1, 2, 3, 4, 5, 6, 7], 6))
print(Solution.place_gas_stations([1, 2, 3, 4, 5], 4))
print(Solution.place_gas_stations(list(range(1, 11)), 1))
print(Solution.place_gas_stations([3, 6, 12, 19, 33, 44, 67, 72, 89, 95], 2))
print(Solution.place_gas_stations([1, 13, 17, 23], 5))
